File: file.cpp
#include <bits/stdc++.h>
using namespace std;
 
class Student
{
public:
    int sID, sAge;
    char sName[10], sGrade[10];
};
 
class Tracher
{
public:
    int tID, tAge;
    char tName[10], tSubject[10];
};
int ar[1000]{};
int arr[1000]{};
void write_Student()
{
    fstream outFile("Student.txt", ios::out | ios::app);
 
    Student s1;
    char c = 'y';
    while (c == 'y')
    {
        cout << "Enter Student Name : ";
        cin >> s1.sName;
 
        cout << "Enter Student ID : ";
        cin >> s1.sID;
        arr[s1.sID]++;
        while(arr[s1.sID] != 1){
            cout << "this id is valid, input another id";
            cin >> s1.sID;
            arr[s1.sID]++;
            }

        cout << "Enter Student Age : ";
        cin >> s1.sAge;
 
        cout << "Enter Student Grade : ";
        cin >> s1.sGrade;
 
        outFile.write((char *)&s1, sizeof(s1));
        cout << "Enter records again? (y/n) ";
        cin >> c;
    }
    outFile.close();
}
 
 
void write_Teacher()
{
    fstream outFile("Teacher.txt", ios::out | ios::app);
    Tracher t1;
    char c = 'y';
    while (c == 'y')
    {
        cout << "Enter Teacher Name : ";
        cin >> t1.tName;
 
        cout << "Enter Teacher ID : ";
        cin >> t1.tID;
        arr[t1.tID]++;
        while(arr[t1.tID] != 1){
            cout << "this id is valid, input another id";
            cin >> t1.tID;
            arr[t1.tID]++;
            }
 
        cout << "Enter Teacher Age : ";
        cin >> t1.tAge;
 
        cout << "Enter Teacher Subject : ";
        cin >> t1.tSubject;
 
        outFile.write((char *)&t1, sizeof(t1));
        cout << "Enter records again? (y/n) ";
        cin >> c;
    }
    outFile.close();
}
 
void read_Student()
{
    Student s1;
    ifstream in;
    in.open("Student.txt", ios::in);
    if (in.is_open())
    {
        cout << "ID\tName\tAge\tGrade" << endl;
        while (true)
        {
            in.read((char *)&s1, sizeof(s1));
            if (in.eof())
             break;
            cout << s1.sID << "\t" << s1.sName << "\t" << s1.sAge << "\t" << s1.sGrade << endl;
        }
    }
    else
        cout << "This File is not Valid \n";
    in.close();
}
 
void read_Teacher()
{
    Tracher t1;
    ifstream in;
    in.open("Teacher.txt", ios::in);
    if (in.is_open())
    {
        cout << "ID\tName\tAge\tSubject" << endl;
        while (true)
        {
            in.read((char *)&t1, sizeof(t1));
            if (in.eof())
                break;
 
            cout << t1.tID << "\t" << t1.tName << "\t" << t1.tAge << "\t" << t1.tSubject << endl;
        }
    }
    else
        cout << "This File is not Valid \n";
    in.close();
}
 
 
 
void Search_Student(int id)
{
    bool found;
    Student s1;
    ifstream in;
    in.open("Student.txt", ios::in);
    if (in.is_open())
    {
        in.read((char *)&s1, sizeof(s1));
        while (!in.eof())
        {
            if (s1.sID == id)
            {
                cout << "ID\tName\tAge\tGrade" << endl;
                cout << s1.sID << "\t" << s1.sName << "\t" << s1.sAge << "\t" << s1.sGrade << endl;
                found = true;
                break;
            }
            in.read((char *)&s1, sizeof(s1));
        }
        if (!found)
            cout << "ID Not Found" << endl;
 
        in.close();
    }
    else
    {
        cout << "File not found" << endl;
    }
}
 
 
 
void Search_Teacher(int id)
{
    bool found;
    Tracher t1;
    ifstream in;
    in.open("Teacher.txt", ios::in);
    if (in.is_open())
    {
        in.read((char *)&t1, sizeof(t1));
        while (!in.eof())
        {
            if (t1.tID == id)
            {
                cout << "ID\tName\tAge\tSubject" << endl;
                cout << t1.tID << "\t" << t1.tName << "\t" << t1.tAge << "\t" << t1.tSubject << endl;
                found = true;
                break;
            }
            in.read((char *)&t1, sizeof(t1));
        }
        if (!found)
            cout << "ID Not Found" << endl;
 
        in.close();
    }
    else
    {
        cout << "File Not Found" << endl;
    }
}
 
 
 
 
void update_teacher(){
    read_Teacher();
    char str[10];
    cout << "Enter the name of the teacher whose data you want to update :"<<endl;
    cin >> str;
    bool found=false;
    Tracher t;
    fstream f;
    f.open("Teacher.txt",ios::in | ios::out);
    if(f.is_open()){
 
        f.read((char*)&t,sizeof(t));
        while(!f.eof()){
            if(strcmp(str,t.tName)==0){
                cout<<"id"<<"\t"<<"Name"<<"\t"<<"Age"<<"\t"<<"Subject"<<endl;
                cout<<t.tID<<"\t"<<t.tName<<"\t"<<t.tAge<<"\t"<<t.tSubject<<endl;
 
                cout << " 1-> Update teacher's name.\n";
                cout << " 2-> Update teacher's age.\n";
                cout << " 3-> Update teacher's id.\n";
                cout << " 4-> Update teacher's subject.\n";
                int operation;
                cin >> operation;
                switch (operation)
                {
                    case 1:                     
                        cout << "enter the new name for the teacher "<<str<< " \n";
                        cin >> t.tName;
 
                        break;
                    case 2:
 
                        cout << "enter the new age for the teacher "<<str<< " \n";
                        cin >> t.tAge;
 
                        break;
                    case 3:                     
                        cout << "enter the new id for the teacher "<<str<< " \n";
                        cin >> t.tID;
                        break;
                    case 4:                     
                        cout << "enter the new subject for the teacher "<<str<< " \n";
                        cin >> t.tSubject;
                        break;
                }
                int curPos =f.tellg();
                int stuSize=sizeof(t);
                f.seekp(curPos-stuSize,ios::beg);
                f.write((char*)&t,sizeof(t));
                f.seekp(curPos-stuSize,ios::beg);
                f.read((char*)&t,sizeof(t));
                cout<<t.tID<<"\t"<<t.tName<<"\t"<<t.tAge<<"\t"<<t.tSubject<<endl;
                found = true;
                break;
            }
            f.read((char*)&t,sizeof(t));
        }if(!found){
            cout<<"name not found";
        }
 
    f.close();
    }
}
 
 
void update_student(){
    read_Student();
    char str[10];
    cout<<"Enter the name of the student whose data you want to update  "<<endl;
    cin>>str;
    bool found=false;
    Student s;
    fstream f;
    f.open("Student.txt",ios::in | ios::out);
    if(f.is_open()){
 
        f.read((char*)&s,sizeof(s));
        while(!f.eof()){
        /*ask youssef about it*/    if(strcmp(str,s.sName)==0){
                cout<<"id"<<"\t"<<"Name"<<"\t"<<"Age"<<"\t"<<"Grade"<<endl;
                cout<<s.sID<<"\t"<<s.sName<<"\t"<<s.sAge<<"\t"<<s.sGrade<<endl;
 
                cout << " 1-> Update student's name.\n";
                cout << " 2-> Update student's age.\n";
                cout << " 3-> Update student's id.\n";
                cout << " 4-> Update student's grade.\n";
                int operation;
                cin >> operation;
                switch (operation)
                {
                    case 1:                     
                        cout << "enter the new name for the student "<<str<< " \n";
                        cin >> s.sName;
 
                        break;
                    case 2:
 
                        cout << "enter the new age for the student "<<str<< " \n";
                        cin >> s.sAge;
 
                        break;
                    case 3:                     
                        cout << "enter the new id for the student "<<str<< " \n";
                        cin >> s.sID;
                        break;
                    case 4:                     
                        cout << "enter the new grade for the student "<<str<< " \n";
                        cin >> s.sGrade;
                        break;
                }
                int curPos =f.tellg();
                int stuSize=sizeof(s);
                f.seekp(curPos-stuSize,ios::beg);
                f.write((char*)&s,sizeof(s));
                f.seekp(curPos-stuSize,ios::beg);
                f.read((char*)&s,sizeof(s));
                cout<<s.sID<<"\t"<<s.sName<<"\t"<<s.sAge<<"\t"<<s.sGrade<<endl;
                found = true;
                break;
            }
            f.read((char*)&s,sizeof(s));
        }if(!found){
            cout<<"name not found";
        }
 
    f.close();
    }
}
 
void delete_Teacher(){
    int id ;
 
cin>>id;
    bool found;
    Tracher t1;
    ifstream in;
    ofstream Temp("temp.txt",ios::out);
    in.open("Teacher.txt",ios::in);
    if(in.is_open())
    {
        in.read((char*)&t1,sizeof(t1));
        while(!in.eof())
        {
            if( t1.tID != id)
            {
               Temp.write((char*)&t1,sizeof(t1));
                found = true;
            }
            in.read((char*)&t1,sizeof(t1));           
        }
        if(!found)
            cout<<"ID not found";
 
        in.close();
        Temp.close();
        remove("Teacher.txt");
        rename("temp.txt","Teacher.txt");
    }
    else 
       cout<<"ID not found";
    
}
 
 
void delete_student(){
    int id ;
cin>>id;
    bool found;
    Student s1;
    ifstream in;
    ofstream Temp("temp.txt",ios::out);
    in.open("Student.txt" , ios::in);
    if(in.is_open())
    {
        in.read((char*)&s1,sizeof(s1));
        while(!in.eof())
        {
            if(s1.sID != id)
            {
             
          Temp.write((char*)&s1,sizeof(s1));
                found = true;
            }
            in.read((char*)&s1,sizeof(s1));           
        }
        if(!found)
            cout<<"ID not found";
 
        in.close();
        Temp.close();
        remove("Student.txt");
        rename("temp.txt","Student.txt");
    }
    else 
       cout<<"ID not found";
    
}
 
int main()
{
    char c;
    int id = 0;
    int id2 = 0 ;
    while (true)
    {
        // --------------------------Menu--------------------------------------
 
        cout << " 1-> Write data.\n";
        cout << " 2-> Read data.\n";
        cout << " 3-> Search for data.\n";
        cout << " 4-> update for data.\n";
        cout << " 5-> delete for data.\n";
 
 
 
 
        // ---------------------------Select Operation-------------------------------------
 
        int op;
        cin >> op;
 
        switch (op)
        {
            int n;
 
        case 1:
            cout << " 1-> Write Student.\n";
            cout << " 2-> Write Teacher.\n";
            cin >> n;
            switch (n)
            {
            case 1:
                write_Student();
                break;
            case 2:
                write_Teacher();
                break;
            }
            break;
        case 2:
            cout << " 1-> Read Student.\n";
            cout << " 2-> Read Teacher.\n";
            cin >> n;
            switch (n)
            {
            case 1:
                read_Student();
                break;
            case 2:
                read_Teacher();
                break;
            }
            break;
        case 3:
            cout << " 1-> Search For Student.\n";
            cout << " 2-> Search For Teacher.\n";
            cin >> n;
            switch (n)
            {
            case 1:
                cout << "Enter Student ID to Search for : " << endl;
                cin >> id;
                Search_Student(id);
                break;
            case 2:
                cout << "Enter Teacher ID to Search for : " << endl;
                cin >> id;
                Search_Teacher(id);
                break;
            }
            break;
 
            case 4 : 
            cout << " 1-> update For Student.\n";
            cout << " 2-> update For Teacher.\n";
            cin >> n;
            switch (n)
            {
            case 1:
                cout << "Enter Student ID to update for : " << endl;
                update_student();
                break;
            case 2:
                cout << "Enter Teacher ID to update for : " << endl;
                update_teacher();
                break;
            }
            break;
 
            case 5 : 
            cout << " 1-> delete For Student.\n";
            cout << " 2-> delete For Teacher.\n";
            cin >> n;
            switch (n)
            {
            case 1:
                cout << "Enter Student ID to delete for : " << endl;
                delete_student();
                break;
            case 2:
                cout << "Enter Teacher ID to delete for : " << endl;
                delete_Teacher();
                break;
            }
            break;
        }

        cout << "Do you want to do another Operation ? [y/n] -> ";
        cin >> c;
        if (c != 'y')
            break;
    }
}

File: file.py
def writestudent():
    with open("students.txt","r") as file:
 
        ids = [line.split()[0] for line in file]
 
    with open("students.txt","a") as file:
        c = 'y'
        while c == 'y':
            id = input("enter student's id : ")
 
            while id in ids:
                print("ID already exists. Please enter a different ID.")
                id = input("enter student's id : ")
            name = input("enter student's name : ")
            age = input("enter student's age : ")
            grade = input("enter student's grade : ")
            file.write(id + '\t' + name + '\t' + age + '\t' + grade + '\n')
            ids.append(id) 
            c = input("add another student ? (y/n)")
        print("record added successfully.")
#-------------------------------------------------------------
 
def writeteacher():
    with open("teachers.txt","r") as file:
 
        ids = [line.split()[0] for line in file]
 
    with open("teachers.txt","a") as file:
        c = 'y'
        while c == 'y':
            id = input("enter teachar id : ")
 
            while id in ids:
                print("ID already exists. Please enter a different ID.")
                id = input("enter teachar's id : ")
            name = input("enter teachar's name : ")
            age = input("enter teachar's age : ")
            subject = input("enter teachar's subject : ")
            file.write(id + '\t' + name + '\t' + age + '\t' + subject + '\n')
            ids.append(id) 
            c = input("add another teacher ? (y/n)")
        print("record added successfully.")

#-------------------------------------------------------------

def readstudent():
    with open("students.txt","r") as file:
        print("ID" + "\t" + "NAME" + "\t" + "AGE" + "\t" + "GRADE")
        print("---------------------------------------------------")
        for line in file:
            print(line, end='')

#-------------------------------------------------------------

def readteacher():
    with open("teachers.txt","r") as file:
        print("ID" + "\t" + "NAME" + "\t" + "AGE" + "\t" + "SUBJECT")
        print("---------------------------------------------------")
        for line in file:
            print(line, end='')

#-------------------------------------------------------------

def searchstudent():
    with open('students.txt', 'r') as file :
        id = input('Enter the id of the student to search for : ')
        flag = False
        for line in file:
            std = line.split('\t')
            if std [0] == id:
                flag = True
                print("ID" + "\t" + "NAME" + "\t" + "AGE" + "\t" + "GRADE")
                print('------------------------------------------------')
                print(line)
        if not flag:
            print('there is no student found with id : ' + id)

#-------------------------------------------------------------

def searchteacher():
    with open("teachers.txt","r") as file:
        id = input('Enter the id of the teacher to search for : ')
        flag = False
        for line in file:
            tch = line.split('\t')
            if tch [0] == id:
                flag = True
                print("ID" + "\t" + "NAME" + "\t" + "AGE" + "\t" + "SUBJECT")
                print('------------------------------------------------')
                print(line)
        if not flag:
            print('there is no teacher found with id : ' + id)

#-------------------------------------------------------------

def deletestudent():
    import os
    id = input ("Enter the id of student you want to delete: ")
    file = open ("students.txt", "r")
    tempFile = open ("tempstudents.txt", "w")
    flag = False
    for line in file:
        st = line.split ("\t")
        if st[0] == id:
            flag = True
        else:
            tempFile.write (line)
    file.close()
    tempFile.close()
    os.remove("students.txt")                                        
    os.rename("tempstudents.txt", "students.txt")    
    if flag:
        print ("student successfully deleted")
    else:
        print ("No students are found with id : " + id)

#-------------------------------------------------------------

def deleteteacher():
    import os
    id = input ("Enter the id of teacher you want to delete: ")
    file = open ("teachers.txt", "r")
    tempFile = open ("tempteachers.txt", "w")
    flag = False
    for line in file:
        st = line.split ("\t")
        if st[0] == id:
            flag = True
        else:
            tempFile.write (line)
    file.close()
    tempFile.close()
    os.remove("teachers.txt")                                        
    os.rename("tempteachers.txt", "teachers.txt")    
    if flag:
        print ("teacher successfully deleted")
    else:
        print ("No teachers are found with id : " + id)
#-------------------------------------------------------------
def updatestudent () :
    #readstudent()
    import os
    id = input ("Enter the id of the student you want to update : ")
    file= open ("students.txt", "r")
    tempFile = open ("tempstudents.txt", "w")
    flag= False
    for line in file:
        st = line.split ("\t")
        if st[0] == id:
            flag = True
            print("1 -> Update student's id.")
            print("2 -> Update student's name.")
            print("3 -> Update student's age.")
            print("4 -> Update student's grade.")
            print("5 -> Update all student's record.")
            op = input("")
            if op == '1':
                id = input ('Enter the new id for the student : ')
                line = id + '\t' + st[1] + '\t' + st [2] + '\t' + st [3]
            elif op == '2':
                name = input ('Enter the new name for the student : ')
                line = st[0] + '\t' + name + '\t' + st [2] + '\t' + st [3]
            elif op == '3':
                age = input ('Enter the new age for the student : ')
                line = st[0] + '\t' + st[1] + '\t' + age + '\t' + st [3]
            elif op == '4':
                grade = input ('Enter the new grade for the student : ')
                line = st[0] + '\t' + st[1] + '\t' + st [2] + '\t' + grade + '\n'    
            elif op == '5':
                id = input ('Enter the new id for the student : ')
                name = input ('Enter the new name for the student : ')
                age = input ('Enter the new age for the student : ')
                grade = input ('Enter the new grade for the student : ')
                line = id + '\t' + name + '\t' + age + '\t' + grade + '\n'    
        tempFile.write (line)
    file.close ()
    tempFile.close ()
    os. remove ("students.txt")          # delete original file
    os. rename ("tempstudents.txt", "students.txt")       # rename temp file
    if flag:
        print ("student successfully updated")
        readstudent()
    else:
        print ("No students are found with id : " + id)
#-------------------------------------------------------------

def updateteacher():
    #readteacher()
    import os
    id = input ("Enter the id of the teacher you want to update : ")
    file= open ("teachers.txt", "r")
    tempFile = open ("tempteachers.txt", "w")
    flag= False
    for line in file:
        tch = line.split ("\t")
        if tch[0] == id:
            flag = True
            print("1 -> Update teacher's id.")
            print("2 -> Update teacher's name.")
            print("3 -> Update teacher's age.")
            print("4 -> Update teacher's subject.")
            print("5 -> Update all teachers's record.")
            op = input("")
            if op == '1':
                id = input ('Enter the new id for the teacher : ')
                line = id + '\t' + tch[1] + '\t' + tch[2] + '\t' + tch[3]
            elif op == '2':
                name = input ('Enter the new name for the teacher : ')
                line = tch[0] + '\t' + name + '\t' + tch[2] + '\t' + tch[3]
            elif op == '3':
                age = input ('Enter the new age for the teacher : ')
                line = tch[0] + '\t' + tch[1] + '\t' + age + '\t' + tch[3]
            elif op == '4':
                subject = input ('Enter the new subject for the teacher : ')
                line = tch[0] + '\t' + tch[1] + '\t' + tch[2] + '\t' + subject + '\n'    
            elif op == '5':
                id = input ('Enter the new id for the teacher : ')
                name = input ('Enter the new name for the teacher : ')
                age = input ('Enter the new age for the teacher : ')
                subject = input ('Enter the new subject for the teacher : ')
                line = id + '\t' + name + '\t' + age + '\t' + subject + '\n'    
        tempFile.write (line)
    file.close ()
    tempFile.close ()
    os. remove ("teachers.txt")          # delete original file
    os. rename ("tempteachers.txt", "teachers.txt")       # rename temp file
    if flag:
        print ("teacher successfully updated")
        readteacher()
    else:
        print ("No teachers are found with id : " + id)
#-------------------------------------------------------------

def main():
    c = 'y'
    while c == 'y':
        print("1 -> To add data")
        print("2 -> To read data")
        print("3 -> To search data")
        print("4 -> To delete data")
        print("5 -> To update data")
        op = input("enter : ")
        if op == '1':
            print("1 -> To write student")
            print("2 -> To write teacher")
            op2 = input("")
            if op2 == '1':
                writestudent()
            elif op2 == '2':
                writeteacher()
            else:
                print("invalid input")
        elif op == '2':
            print("1 -> To read student")
            print("2 -> To read teacher")
            op2 = input("")
            if op2 == '1':
                readstudent()
            elif op2 == '2':
                readteacher()
            else:
                print("invalid input")
        elif op == '3':
            print("1 -> To search for student")
            print("2 -> To search for teacher")
            op2 = input("")
            if op2 == '1':
                searchstudent()
            elif op2 == '2':
                searchteacher()
            else:
                print("invalid input")
        elif op == '4':
            print("1 -> To delete student")
            print("2 -> To delete teacher")
            op2 = input("")
            if op2 == '1':
                deletestudent()
            elif op2 == '2':
                deleteteacher()
            else:
                print("invalid input")
        elif op == '5':
            print("1 -> To update student")
            print("2 -> To update teacher")
            op2 = input("")
            if op2 == '1':
                updatestudent()
            elif op2 == '2':
                updateteacher()
            else:
                print("invalid input")
        c = input("perform another operation ? (y/n)")
    print("bye bye :)")

main()
